File: python/semantic/small_step.py
# coding: utf-8


class Number(object):
    def __init__(self, value):
        self.value = value

    def __repr__(self):
        return str(self.value)

    def reducible(self):
        return False


class Add(object):
    def __init__(self, left, right):
        self.left = left
        self.right = right

    def __repr__(self):
        return '{0} + {1}'.format(self.left, self.right)

    def reducible(self):
        return True

    def reduce(self, env):
        if self.left.reducible():
            return Add(self.left.reduce(env), self.right)
        elif self.right.reducible():
            return Add(self.left, self.right.reduce(env))
        else:
            return Number(self.left.value + self.right.value)


class Multiply(object):
    def __init__(self, left, right):
        self.left = left
        self.right = right

    def __repr__(self):
        return '{0} * {1}'.format(self.left, self.right)

    def reducible(self):
        return True

    def reduce(self, env):
        if self.left.reducible():
            return Multiply(self.left.reduce(env), self.right)
        elif self.right.reducible():
            return Multiply(self.left, self.right.reduce(env))
        else:
            return Number(self.left.value * self.right.value)


class Boolean(object):
    def __init__(self, value):
        self.value = value

    def __repr__(self):
        return str(self.value)

    def reducible(self):
        return False


class LessThan(object):
    def __init__(self, left, right):
        self.left = left
        self.right = right

    def __repr__(self):
        return '{0} < {1}'.format(self.left, self.right)

    def reducible(self):
        return True

    def reduce(self, env):
        if self.left.reducible():
            return LessThan(self.left.reduce(env), self.right)
        elif self.right.reducible():
            return LessThan(self.left, self.right.reduce(env))
        else:
            return Boolean(self.left.value < self.right.value)


class Variable(object):
    def __init__(self, name):
        self.name = name

    def __repr__(self):
        return str(self.name)

    def reducible(self):
        return True

    def reduce(self, env):
        return env[self.name]


class DoNothing(object):
    def __eq__(self, other_statement):
        return isinstance(other_statement, DoNothing)

    def __repr__(self):
        return 'do_nothing'

    def reducible(self):
        return False


class Assign(object):
    def __init__(self, name, expr):
        self.name = name
        self.expr = expr

    def __repr__(self):
        return '{0} = {1}'.format(self.name, self.expr)

    def reducible(self):
        return True

    def reduce(self, env):
        if self.expr.reducible():
            return Assign(self.name, self.expr.reduce(env)), env
        else:
            env.update({self.name: self.expr})
            return DoNothing(), env


class If(object):
    def __init__(self, condition, sequence, alternative):
        self.condition = condition
        self.sequence = sequence
        self.alternative = alternative

    def __repr__(self):
        return 'If ({0}) {{ {1} }} else {{ {2} }}'.format(
            self.condition, self.sequence, self.alternative
        )

    def reducible(self):
        return True

    def reduce(self, env):
        if self.condition.reducible():
            return [If(self.condition.reduce(env), self.sequence, self.alternative), env]
        else:
            if self.condition.value == Boolean(True).value:
                return [self.sequence, env]
            else:
                return [self.alternative, env]


class Sequence(object):
    def __init__(self, first, second):
        self.first = first
        self.second = second

    def __repr__(self):
        return '{0}; {1}'.format(self.first, self.second)

    def reducible(self):
        return True

    def reduce(self, env):
        if isinstance(self.first, DoNothing):
            return self.second, env
        first, env = self.first.reduce(env)
        return Sequence(first, self.second), env


class While(object):
    def __init__(self, condition, body):
        self.condition = condition
        self.body = body

    def __repr__(self):
        return 'while ({0}) {{ {1} }}'.format(self.condition, self.body)

    def reducible(self):
        return True

    def reduce(self, env):
        return If(self.condition, Sequence(self.body, self), DoNothing()), env


class Machine(object):
    def __init__(self, statement, env):
        self.statement = statement
        self.env = env

    def step(self):
        self.statement, self.env = self.statement.reduce(self.env)

    def run(self):
        while self.statement.reducible():
            print '{0}, {1}'.format(self.statement, self.env)
            self.step()
        print '{0}, {1}'.format(self.statement, self.env)


## Test "If", "Sequence", "Assign"
# statement = If(Variable('x'),
#                Sequence(Assign('y', Number(1)),
#                         Assign('z', Number(2))),
#                DoNothing())
# env = {'x': Boolean(True)}

## Test "While"
statement = While(LessThan(Variable('x'), Number(6)),
                  Assign('x', Multiply(Variable('x'), Number(2))))
env = {'x': Number(1)}

Machine(statement, env).run()

File: ruby/semantic/small_step.rb


class Number < Struct.new(:value)
  def to_s
    "#{value}"
  end

  def inspect
    "<<#{self}>>"
  end

  def reducible?
    false
  end
end


class Add < Struct.new(:left, :right)
  def to_s
    "#{left}+#{right}"
  end

  def reducible?
    true
  end

  def reduce(env)
    if left.reducible?
      Add.new(left.reduce(env), right)
    elsif right.reducible?
      Add.new(left, right.reduce(env))
    else
      Number.new(left.value + right.value)
    end
  end
end


class Multiply < Struct.new(:left, :right)
  def to_s
    "#{left}*#{right}"
  end

  def reducible?
    true
  end

  def reduce(env)
    if left.reducible?
      Multiply.new(left.reduce(env), right)
    elsif right.reducible?
      Multiply.new(left, right.reduce(env))
    else
      Number.new(left.value * right.value)
    end
  end
end


class Boolean < Struct.new(:value)
  def to_s
    value.to_s
  end

  def inspect
    "<<#{self}>>"
  end

  def reducible?
    false
  end
end


class LessThan < Struct.new(:left, :right)
  def to_s
    "#{left} < #{right}"
  end

  def reducible?
    true
  end

  def reduce(env)
    if left.reducible?
      LessThan.new(left.reduce(env), right)
    elsif right.reducible?
      LessThan.new(left, right.reduce(env))
    else
      Boolean.new(left.value < right.value)
    end
  end
end


class Variable < Struct.new(:name)
  def to_s
    name.to_s
  end

  def reducible?
    true
  end

  def reduce(env)
    env[name]
  end
end


class DoNothing
  def to_s
    'do_nothing'
  end

  def ==(other_statement)
    other_statement.instance_of?(DoNothing)
  end

  def reducible?
    false
  end
end


class Assign < Struct.new(:name, :expr)
  def to_s
    "#{name} = #{expr}"
  end

  def reducible?
    true
  end

  def reduce(env)
    if expr.reducible?
      [Assign.new(name, expr.reduce(env)), env]
    else
      [DoNothing.new, env.merge({name => expr})]
    end
  end
end


class If < Struct.new(:condition, :consequence, :alternative)
  def to_s
    "if (#{condition}) { #{consequence} } else { #{alternative} }"
  end

  def reducible?
    true
  end

  def reduce(env)
    if condition.reducible?
      [If.new(condition.reduce(env), consequence, alternative), env]
    else
      case condition
        when Boolean.new(true)
          [consequence, env]
        when Boolean.new(false)
          [alternative, env]
      end
    end
  end
end


class Sequence < Struct.new(:first, :second)
  def to_s
    "#{first}; #{second}"
  end

  def reducible?
    true
  end

  def reduce(env)
    case first
      when DoNothing.new
        [second, env]
      else
        reduced_first, reduced_env = first.reduce(env)
        [Sequence.new(reduced_first, second), reduced_env]
    end
  end
end


class While < Struct.new(:condition, :body)
  def to_s
    "while (#{condition}) { #{body} }"
  end

  def reducible?
    true
  end

  def reduce(env)
    [If.new(condition, Sequence.new(body, self), DoNothing.new), env]
  end
end


class Machine < Struct.new(:statement, :env)
  def step
    self.statement, self.env = statement.reduce(env)
  end

  def run
    while statement.reducible?
      puts "#{statement}, #{env}"
      step
    end

    puts "#{statement}, #{env}"
  end
end


# Test "If", "Sequence", "Assign"
statement = If.new(Variable.new(:x),
                   Sequence.new(Assign.new(:y, Number.new(1)),
                                Assign.new(:z, Number.new(2))),
                   DoNothing.new)
env = {x: Boolean.new(true)}
Machine.new(statement, env).run

# Test "While"
statement = While.new(
                     LessThan.new(Variable.new(:x), Number.new(6)),
                     Assign.new(:x, Multiply.new(Variable.new(:x), Number.new(2)))
                     )
env = {x: Number.new(1)}

Machine.new(statement, env).run
